experiment/needs-rebase/main.go

```diff
@@ -0,0 +1,164 @@
+	// TODO: Use global option from the prow config.
+	logrus.SetLevel(logrus.WarnLevel)
+	log := logrus.StandardLogger().WithField("plugin", "needs-rebase")
+
+	// Ignore SIGTERM so that we don't drop hooks when the pod is removed.
+	// We'll get SIGTERM first and then SIGKILL after our graceful termination
+	// deadline.
+	signal.Ignore(syscall.SIGTERM)
+
+	webhookSecretRaw, err := ioutil.ReadFile(*webhookSecretFile)
+	if err != nil {
+		log.WithError(err).Fatal("Could not read webhook secret file.")
+	}
+	webhookSecret := bytes.TrimSpace(webhookSecretRaw)
+
+	oauthSecretRaw, err := ioutil.ReadFile(*githubTokenFile)
+	if err != nil {
+		log.WithError(err).Fatal("Could not read oauth secret file.")
+	}
+	oauthSecret := string(bytes.TrimSpace(oauthSecretRaw))
+
+	_, err = url.Parse(*githubEndpoint)
+	if err != nil {
+		log.WithError(err).Fatal("Must specify a valid --github-endpoint URL.")
+	}
+
+	pa := &plugins.PluginAgent{}
+	if err := pa.Start(*pluginConfig); err != nil {
+		log.WithError(err).Fatalf("Error loading plugin config from %q.", *pluginConfig)
+	}
+
+	githubClient := github.NewClient(oauthSecret, *githubEndpoint)
+	if *dryRun {
+		githubClient = github.NewDryRunClient(oauthSecret, *githubEndpoint)
+	}
+	githubClient.Throttle(360, 360)
+
+	server := &Server{
+		hmacSecret: webhookSecret,
+
+		ghc: githubClient,
+		log: log,
+	}
+
+	go periodicUpdate(log, pa, githubClient, *updatePeriod)
+
+	http.Handle("/", server)
+	externalplugins.ServeExternalPluginHelp(http.DefaultServeMux, log, plugin.HelpProvider)
+	logrus.Fatal(http.ListenAndServe(":"+strconv.Itoa(*port), nil))
+}
+
+// Server implements http.Handler. It validates incoming GitHub webhooks and
+// then dispatches them to the appropriate plugins.
+type Server struct {
+	hmacSecret []byte
+
+	ghc *github.Client
+	log *logrus.Entry
+}
+
+// ServeHTTP validates an incoming webhook and puts it into the event channel.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// TODO: Move webhook handling logic out of hook binary so that we don't have to import all
+	// plugins just to validate the webhook.
+	eventType, eventGUID, payload, ok := hook.ValidateWebhook(w, r, s.hmacSecret)
+	if !ok {
+		return
+	}
+	fmt.Fprint(w, "Event received. Have a nice day.")
+
+	if err := s.handleEvent(eventType, eventGUID, payload); err != nil {
+		logrus.WithError(err).Error("Error parsing event.")
+	}
+}
+
+func (s *Server) handleEvent(eventType, eventGUID string, payload []byte) error {
+	l := s.log.WithFields(
+		logrus.Fields{
+			"event-type":     eventType,
+			github.EventGUID: eventGUID,
+		},
+	)
+	switch eventType {
+	case "pull_request":
+		var pre github.PullRequestEvent
+		if err := json.Unmarshal(payload, &pre); err != nil {
+			return err
+		}
+		go func() {
+			if err := plugin.HandleEvent(l, s.ghc, &pre); err != nil {
+				l.Info("Error handling event.")
+			}
+		}()
+	default:
+		return fmt.Errorf("received an event of type %q but didn't ask for it", eventType)
```

Comment:
This was very annoying when I started running the cherrypick plugin as a standalone server so I ended up logging it in Debug mode.